### ISSUE TITLE
Document LearningRatePolicy

### ISSUE THREAD

**Issue by nyghtowl** [deeplearning4j/deeplearning4j#1494]
LearningRatePolicy provides decay alternatives during training. Explain how to apply the following:

```
None,
Exponential,
Inverse,
Poly,
Sigmoid,
Step,
Schedule,
Score
```

Use this [test set](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-core/src/test/java/org/deeplearning4j/nn/updater/TestDecayPolicies.java) to help with documentation.

**Comment by tomthetrainer** [comment]
# Dynamic Learning Rate Policy

## Contents

* <a href="#description">Description</a>
* <a href="#examples">Examples and Use Cases</a>
* <a href="#setup">Setup</a>
    * <a href="#prereqs">Prerequisites</a>
    * <a href="#step-by-step">Step-by-step</a>
      * <a href="#step-1">Step One</a>
      * <a href="#step-2">Step Two</a>
* <a href="#troubleshooting">Troubleshooting</a>
* <a href="#further-reading">Further reading</a>

## <a name="description">Description</a>

LearningRatePolicy provides decay alternatives during training. 



## <a name="examples">Examples and Use Cases</a>


Exponential,

```
double lr = 1e-2;
double decayRate = 2;
NeuralNetConfiguration conf = new NeuralNetConfiguration.Builder()
.learningRate(lr)
.learningRateDecayPolicy(LearningRatePolicy.Exponential)
.lrPolicyDecayRate(decayRate).iterations(iterations)
.layer(new DenseLayer.Builder().nIn(nIn).nOut(nOut)
.updater(org.deeplearning4j.nn.conf.Updater.SGD).build())
.build();
```
Inverse,

```
double lr = 1e-2;
double decayRate = 2;
double power = 3;
NeuralNetConfiguration conf = new NeuralNetConfiguration.Builder().learningRate(lr)
.learningRateDecayPolicy(LearningRatePolicy.Inverse)
.lrPolicyDecayRate(decayRate).lrPolicyPower(power).iterations(iterations)
.layer(new DenseLayer.Builder().nIn(nIn).nOut(nOut)
.updater(org.deeplearning4j.nn.conf.Updater.SGD).build())
.build();
```

Poly,
```
NeuralNetConfiguration conf = new NeuralNetConfiguration.Builder().learningRate(lr)
.learningRateDecayPolicy(LearningRatePolicy.Poly).lrPolicyPower(power)
.iterations(iterations)
.layer(new DenseLayer.Builder().nIn(nIn).nOut(nOut)    
.updater(org.deeplearning4j.nn.conf.Updater.SGD).build())                                      
.build();
```

Sigmoid,
```
NeuralNetConfiguration conf = new NeuralNetConfiguration.Builder().learningRate(lr)
.learningRateDecayPolicy(LearningRatePolicy.Sigmoid)
.lrPolicyDecayRate(decayRate).lrPolicySteps(steps).iterations(iterations)
.layer(new DenseLayer.Builder().nIn(nIn).nOut(nOut)
.updater(org.deeplearning4j.nn.conf.Updater.SGD).build())                                
.build();
```

Step,

```
double lr = 1e-2;
double decayRate = 2;
double steps = 3;
NeuralNetConfiguration conf = new NeuralNetConfiguration.Builder().learningRate(lr)
.learningRateDecayPolicy(LearningRatePolicy.Step).lrPolicyDecayRate(decayRate)
.lrPolicySteps(steps).iterations(iterations)
.layer(new DenseLayer.Builder().nIn(nIn).nOut(nOut)
.updater(org.deeplearning4j.nn.conf.Updater.SGD).build())
.build();
```
Schedule

Allows you to specify a schedule

```
MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder().optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT).iterations(1)
.learningRateDecayPolicy(LearningRatePolicy.Schedule)                            
.learningRateSchedule(learningRateSchedule)
```

Score

```
MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
.learningRate(lr)
.learningRateDecayPolicy(LearningRatePolicy.Score).lrPolicyDecayRate(lrScoreDecay).list()
```


## <a name="setup">Setup</a>

There is no special setup required to use this feature.

### <a name="prereqs">Prerequisites</a>

Versions 0.7.0 and above support LearningRateDecayPolicy


### <a name="step-by-step">Step By Step Instructions</a>

1. Configure your Neural network to use a LearningRatePolicy
2. Train your Network

## Troubleshooting

It is possible to choose values that actually cause the training to take longer than it would with a standard updater without a policy applied. Verify that your network is making reasonable progress. 

## Further reading

Tutorial on Keras implemetation of similar feature is [here.](http://machinelearningmastery.com/using-learning-rate-schedules-deep-learning-models-python-keras)

**Comment by tomthetrainer** [comment]
@maragordon Perhaps we should create draft in ticket and edit here and publish after complete?

**Comment by maragordon** [comment]
Revised and added some comments in brackets re: what still needs to be filled in.

**Comment by whatbeg** [comment]
It seems there is no "Score" Policy's implementation, because its source code is like "newLr = lr", which keep lr constantly in training.? Is it true?

**Comment by maxpumperla** [comment]
@tomthetrainer @agibsonccc to understand this better: 
- where do we want to put this kind of documentation?
- what is left for this ticket? Do we just want to "fill in the brackets" above and put it on our website?

Just want to understand the scope here.

**Comment by tomthetrainer** [comment]
@maxpumperla 

Overview: this ticket was created as a way to get the docs in a standard format so it was chosen as an example of a new page that can be created in a standard template format. 

You could fill in the [[brackets ]] and publish, or we can loop @bkanui on this and verify we have a reasonable template.

**Comment by bkanui** [comment]
@maxpumperla @tomthetrainer 
Not knowing the actual implementation, this proposed outline/template looks good to me.
re: where to store the actual files. Could we have a meeting to discuss? I agree, the jumping around that occurs from link to link is disconcerting and confusing. A 'simple' reorg of exiting files through the sidebar/top bar html file should not cause too much trouble.  I'll respond to the other email on that.

**Comment by maxpumperla** [comment]
@AlexDBlack this might be outdated a little with your latest changes, no?

**Comment by AlexDBlack** [comment]
Yes, they still need to be documented, but the actual details have changed considerably.
Schedules are now classes, and both iteration and epoch schedules are supported:
https://github.com/deeplearning4j/nd4j/tree/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/schedule

LR schedules are set directly (and only) in the updaters.

**Comment by AlexDBlack** [comment]
https://deeplearning4j.org/quickref#config-schedules
